ta/pkcs11/src/processing_asymm.c

```diff
@@ -0,0 +1,558 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <compiler.h>
+#include <tee_api_defines.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+
+#include "attributes.h"
+#include "pkcs11_helpers.h"
+#include "pkcs11_token.h"
+#include "processing.h"
+#include "serializer.h"
+
+bool processing_is_tee_asymm(uint32_t proc_id)
+{
+	switch (proc_id) {
+	/* EC flavors */
+	case PKCS11_CKM_ECDSA:
+	case PKCS11_CKM_ECDSA_SHA1:
+	case PKCS11_CKM_ECDSA_SHA224:
+	case PKCS11_CKM_ECDSA_SHA256:
+	case PKCS11_CKM_ECDSA_SHA384:
+	case PKCS11_CKM_ECDSA_SHA512:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static enum pkcs11_rc
+pkcs2tee_algorithm(uint32_t *tee_id, uint32_t *tee_hash_id,
+		   enum processing_func function __unused,
+		   struct pkcs11_attribute_head *proc_params,
+		   struct pkcs11_object *obj)
+{
+	static const uint32_t pkcs2tee_algo[][3] = {
+		/* EC flavors (Must find key size from the object) */
+		{ PKCS11_CKM_ECDSA, 1, 0 },
+		{ PKCS11_CKM_ECDSA_SHA1, 1, TEE_ALG_SHA1 },
+		{ PKCS11_CKM_ECDSA_SHA224, 1, TEE_ALG_SHA224 },
+		{ PKCS11_CKM_ECDSA_SHA256, 1, TEE_ALG_SHA256 },
+		{ PKCS11_CKM_ECDSA_SHA384, 1, TEE_ALG_SHA384 },
+		{ PKCS11_CKM_ECDSA_SHA512, 1, TEE_ALG_SHA512 },
+	};
+	size_t end = sizeof(pkcs2tee_algo) / (2 * sizeof(uint32_t));
```

Comment:
oups! maybe crosscheck this:
```patch
-	size_t end = sizeof(pkcs2tee_algo) / (2 * sizeof(uint32_t));
+	size_t end = sizeof(pkcs2tee_algo) / (3 * sizeof(uint32_t));
```

Reply:
Better to use `ARRAY_SIZE()`.

Reply:
`ARRAY_SIZE(pkcs2tee_algo)` is 3 here, no?
using a struct instead of 3 uint32_t cell rows would make it clearer maybe.

Reply:
I hope not... or it could not be used with multi-dimensional arrays :-/ `ARRAY_SIZE(pkcs2tee_algo)` should be 6.

Reply:
Right. Let's fix this.

Reply:
Maybe prefer the way it's done in _processing_symm.c_:
```c
	static const struct {
		enum pkcs11_mechanism_id mech_id;
		uint32_t tee_id;
		uint32_t hash_id;
	} pkcs2tee_algo[] = {
		/* EC flavors (Must find key size from the object) */
		{ PKCS11_CKM_ECDSA, 1, 0 },
		...
	};

Reply:
Updated like it was in processing_symm in different spot... will still change mech -> mech_id.

Reply:
Updated.